feat: add api boundary nodes to state tree (#248)
dfinity/interface-spec: spec/index.md

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary bodes information {#state-tree-subnet}

> The state tree contains information about all API Boundary Nodes, which can communicate with the Internet Computer (replica nodes). The source of truth for these Api boundary node records is stored in the registry canister and can be added/modified/removed via proposals.

Reviewer: ```suggestion
The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (replica nodes). The source of truth for these API boundary node records is stored in the registry canister and can be added/modified/removed via proposals.
```
the other way around, just make it the same for consistency

> - `api_boundary_nodes/<node_id>/domain` (text)

Reviewer 2: ```suggestion
- `/api_boundary_nodes/<node_id>/domain` (text)
```

>     The domain name associated with a node, e.g., `api-bn1.dfinity.org`.

Reviewer: ```suggestion
    The domain name associated with a node, e.g., `api-bn1.example.com`.
```

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary bodes information {#state-tree-subnet}
The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (replica nodes). The source of truth for these API boundary node records is stored in the registry canister and can be added/modified/removed via proposals.
- `/api_boundary_nodes/<node_id>/domain` (text)

> - `/api_boundary_nodes/<node_id>/ipv4_address` (text)

Reviewer 3: The `ipv4_address` is a Option<string> in Rust. Because the interface specifies it as pure text, we should not encode the `None` case (Otherwise, cbor encodes `None` as some special code). 
Instead, we just skip the fork with label `ipv4_address` in the state tree and don't populate it. Is this the expected behavior? @nikolay-komarevskiy @mraszyk

Reviewer 2: I think that's fine and we can add a note similar to the path `/canister/<canister_id>/module_hash` to characterize that the path is sometimes omitted.

Author: @mraszyk add a short comment

> - `/api_boundary_nodes/<node_id>/ipv6_address` (text)

Reviewer 4: Who would actually read the key from here and use it? I think clients would usually use the domain and get the key from the certificate; IC nodes don't need to read it from the state tree.

> - `/api_boundary_nodes/<node_id>/ipv6_address` (text)

Author: Yes, we confirmed with @r-birkner , let's not put it into the tree and into the spec, accordingly.

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary bodes information {#state-tree-subnet}
The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (replica nodes). The source of truth for these API boundary node records is stored in the registry canister and can be added/modified/removed via proposals.
- `/api_boundary_nodes/<node_id>/domain` (text)

> - `/api_boundary_nodes/<node_id>/public_key` (text)

Reviewer: ```suggestion
```
Remove it for now until we are ready to add it.

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary bodes information {#state-tree-subnet}

> The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (replica nodes). The source of truth for these API boundary node records is stored in the registry canister.

Reviewer 2: ```suggestion

The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (its replica nodes). The source of truth for these API boundary node records is stored in the NNS registry canister.

```

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary bodes information {#state-tree-subnet}

The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (its replica nodes). The source of truth for these API boundary node records is stored in the NNS registry canister.

- `/api_boundary_nodes/<node_id>/domain` (text)
- `/api_boundary_nodes/<node_id>/ipv4_address` (text)
  
    If the `ipv4_address` is empty, this path does not exist.
- `/api_boundary_nodes/<node_id>/ipv6_address` (text)

>     The domain name associated with a node, e.g., `api-bn1.example.com`.

Reviewer 2: this seems to belong to the domain path above

    All partial state trees include a timestamp, indicating the time at which the state is current.

> ### Api boundary nodes information {#state-tree-subnet}

Reviewer 2: ```suggestion
### Api boundary nodes information {#state-tree-api-bn}
```

The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (its replica nodes). The source of truth for these API boundary node records is stored in the NNS registry canister.

- `/api_boundary_nodes/<node_id>/domain` (text)

    Domain name associated with a node. All domains are unique across nodes.
    Example: `api-bn1.example.com`.

> - `/api_boundary_nodes/<node_id>/ipv4_address` (text)

Reviewer 2: ```suggestion

- `/api_boundary_nodes/<node_id>/ipv4_address` (text)
```

  
    Public IPv4 address of a node in the dotted-decimal notation.

>     If the `ipv4_address` is empty, this path does not exist.  

Reviewer 2: ```suggestion
    If no `ipv4_address` is available for the corresponding node, then this path does not exist.  
```

    Example: `192.168.10.150`.

> - `/api_boundary_nodes/<node_id>/ipv6_address` (text)

Reviewer 2: ```suggestion

- `/api_boundary_nodes/<node_id>/ipv6_address` (text)
```

    Public IPv6 address of a node in the hexadecimal notation with colons.
    Example: `3002:0bd6:0000:0000:0000:ee00:0033:6778`.

> ### Subnet information {#state-tree-subnet}

Reviewer 2: ```suggestion

### Subnet information {#state-tree-subnet}
```

    All partial state trees include a timestamp, indicating the time at which the state is current.

### Api boundary nodes information {#state-tree-api-bn}

> The state tree contains information about all API boundary nodes. The source of truth for these API boundary node records is stored in the NNS registry canister.

Reviewer 5: ```suggestion
The state tree contains information about all API boundary nodes (the source of truth for these API boundary node records is stored in the NNS registry canister).
```

> The state tree contains information about all API boundary nodes, which can communicate with the Internet Computer (its replica nodes). The source of truth for these API boundary node records is stored in the NNS registry canister.

Reviewer 2: ```suggestion
The state tree contains information about all API boundary nodes. The source of truth for these API boundary node records is stored in the NNS registry canister.
```

All requested paths must have the following form:

-   `/time`. Can always be requested.

> - `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`. Can always be requested.

Reviewer 2: ```suggestion
- `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`. Can always be requested.

```

To be sure the list is interpreted properly by Markdown.

All requested paths must have the following form:

-   `/time`. Can always be requested.

> - `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`, `/api_boundary_nodes/<node_id>/pub_key`, `/api_boundary_nodes/<node_id>/ip4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.

Reviewer 3: Two questions:
1. Is the api_boundary_nodes information replicated in the state tree of all subnets or just in root subnet?
2. Is the request sent to /api/v2/canister/<effective_canister_id>/read_state`  or `/api/v2/subnet/<subnet_id>/read_state`? And which effective_canister_id/subnet_id? This will decide which subnet handles the request.

Reviewer 2: I believe the information should be replicated across subnets for the sake of load balancing.

Reviewer 3: Please see the [slides](https://docs.google.com/presentation/d/1TyHR0uRiHZsXSTWsuMjnlWJ5DP0hu8C6afaGXGDKMjE/edit#slide=id.g29b6a5284b2_0_523) for the visualization of depth calculation for `/api_boundary_nodes`
Here are the results:

- `/api_boundary_nodes` make `/subnet` one more layer deep

- In mainnet topology,  the depth of `/subnet` increases from 22 to 23

- `/api_boundary_nodes` has the max depth of 15 if there are fewer than 64 api boundary nodes

> - `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`, `/api_boundary_nodes/<node_id>/pub_key`, `/api_boundary_nodes/<node_id>/ip4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.

Reviewer: ```suggestion
- `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`, `/api_boundary_nodes/<node_id>/pub_key`, `/api_boundary_nodes/<node_id>/ipv4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.
```

> - `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`, `/api_boundary_nodes/<node_id>/pub_key`, `/api_boundary_nodes/<node_id>/ipv4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.

Reviewer: ```suggestion
- `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`,  `/api_boundary_nodes/<node_id>/ipv4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.
```
Remove the public key.

All requested paths must have the following form:

-   `/time`. Can always be requested.

> - `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`,  `/api_boundary_nodes/<node_id>/ipv4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.

Reviewer 2: ```suggestion

- `/api_boundary_nodes`, `/api_boundary_nodes/<node_id>`, `/api_boundary_nodes/<node_id>/domain`,  `/api_boundary_nodes/<node_id>/ipv4_address`, `/api_boundary_nodes/<node_id>/ipv6_address`. Can always be requested.
```

Reviewer 2: adding a new line to make the list rendered properly
